### ISSUE TITLE
Response header set via `reply.header()` in a custom handler never gets sent

### ISSUE THREAD

**Issue by rgrove** [fastify/fastify-nextjs#84]
## 🐛 Bug Report

A response header that's set via `reply.header()` or `reply.headers()` in a custom handler will never be sent.

This is because Fastify buffers headers internally instead of setting them on the `http.ServerResponse` instance, and only sends them after `reply.send()` is called. But since Next.js calls `.end()` directly on the `http.ServerResponse` instance, Fastify never has an opportunity to set response headers, and the buffered headers never get sent.

One way to address this might be for fastify-nextjs to iterate over Fastify's response headers and call `reply.res.setHeader(...)` to set each header on the `http.ServerResponse` before passing control to the custom handler or Next.js. Unfortunately, Fastify doesn't appear to provide any public API to get all the internally buffered headers; the only option is to get an individual header by name via `reply.getHeader()`, which is insufficient. I'm curious if anyone has ideas.

## To Reproduce

Here's a failing test that reproduces this:

```js
test('should serve Fastify response headers set by a custom handler', t => {
  t.plan(3)

  const fastify = Fastify()

  fastify
    .register(require('./index'))
    .after(() => {
      fastify.next('/hello', (app, req, reply) => {
        reply.header('test-header', 'hello');
        app.render(req.req, reply.res, '/hello', req.query, {})
      })
    })

  fastify.inject({
    url: '/hello',
    method: 'GET'
  }, (err, res) => {
    t.error(err)
    t.equal(res.statusCode, 200)
    t.equal(res.headers['test-header'], 'hello')
  })

  fastify.close()
});
```

## Expected behavior

The response served by Next.js should include headers set by Fastify, and the test above should pass.

## Your Environment

- *node version*: 12.15.0
- *fastify version*: 2.12.1
- *os*: Mac, Linux

**Comment by mcollina** [comment]
Would you like to send a PR? We also need somebody to help in maintaining this module.

**Comment by rgrove** [comment]
Yep, I'd be happy to send a PR, but I'm afraid I don't see a good way to fix this without upstream changes to Fastify itself, since Fastify's public API doesn't currently provide a way to get all response headers.

If you think Fastify would be open to a PR that adds something like a `reply.getHeaders()` method, then I'd be happy to do that too. But I thought I'd bring it up here first to see if anyone had a better idea.

**Comment by mcollina** [comment]
That'd be nice, yes!

**Comment by rqbazan** [comment]
Hey, @mcollina ✋ Recently I've tried to migrate my custom NextJS server from Express to Fastify, and in my view, what @rgrove said is a limitation of Fastify

> This is because Fastify buffers headers internally instead of setting them on the http.ServerResponse instance, and only sends them after reply.send() is called. But since Next.js calls .end() directly on the http.ServerResponse instance.

`getHeaders` represents a good patch for this specific case, but the whole point could be fixed if Fastify extends [from ServerResponse](https://github.com/expressjs/express/blob/3ff1dbeb73732b6aae73f8df407e3bb423ccc3b7/lib/response.js#L42) and also [from IncomingMessage](https://github.com/expressjs/express/blob/3ff1dbeb73732b6aae73f8df407e3bb423ccc3b7/lib/request.js#L31) like Express does

So, when an isomorphic application built with NextJS tries to use `req` or `res` in the server-side, it'll use the `req` and `reply` of Fastify, not the raw objects. Hence, all the enhancements like the parsed body, cookies, parsed query params, reply decorations, request decorations, etc are preserved. 😃 

```diff
@@ -43,7 +43,7 @@ function fastifyNext (fastify, options, next) {
         return callback(app, req, reply)
       }

-      return handleNextRequests(req.raw, reply.raw)
+      return handleNextRequests(req, reply)
     }
   }
 }
```

**Comment by mcollina** [comment]
There are quite a few reasons to not do what express does, for completeness, I'm listing a few here:

1. `IncomingMessage` and `ServerResponse` are extremely tied to the HTTP/1 implementation of Node.js core. Which make supporting http/2 and in the future http/3 impossible for Express.
2. Avoid monkeypatching - in order to work, Express monkeypatches the internals of those instances, fundamentally changing how they work. This makes it brittle and complex to maintain over time. Fastify relies only on public methods.
3. Encapsulation and performance. A large amount of the performance gap between Express and Fastify can be attributed to the use of our won custom and `Request` and `Reply` instances.

**Comment by rqbazan** [comment]
I got it @mcollina, so sadly there is no way to use `reply` and `request` of Fastify in a NextJS app (`getServerSideProps`), and based on the 2nd point, I shouldn't try to use it in the first place. 😕

**Comment by franky47** [comment]
Sorry for necroposting, but I was hitting a similar issue as @rgrove, where I was trying to set cookies on the Fastify reply in `getServerSideProps` (by passing the Fastify reply as a custom property on the raw request). I'm posting my solution here as it may help others.

The fix in #112 only applies headers that have been set **before** running the Next.js handler, and in my case the `Set-Cookie` header is set **within** the handler. This requires to copy any headers set on the Fastify reply back to the raw response before Next.js sends it. Since the send part is done within Next.js code, there is not much this plugin can do after the fact, so the copy has to happen within `getServerSideProps`:

```ts
// Next.js plugin
fastify.next('/server-side-route', {
  onRequest(request, reply, done) {
    // inject custom properties
    request.raw.fastifyReply = reply
  }
})
```

```ts
// pages/server-side-route.tsx
const getServerSideProps()  = async ctx => {
  ctx.req.fastifyReply.header('x-foo', 'bar')
  ctx.res.setHeader('x-foo', ctx.req.fastifyReply.getHeader('x-bar')!)
}
```
